File: src/minecraft/net/machinemuse/general/sound/Musique.scala
package net.machinemuse.general.sound

import cpw.mods.fml.common.FMLCommonHandler
import cpw.mods.fml.relauncher.Side
import net.minecraft.client.Minecraft
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.client.audio.SoundManager
import scala.Predef.String
import net.machinemuse.general.MuseLogger
import net.machinemuse.powersuits.common.Config

/**
 * Handles sound mechanics
 */
object Musique {
  def soundsystem = SoundManager.sndSystem

  def mcsound = Minecraft.getMinecraft.sndManager

  def options = Minecraft.getMinecraft.gameSettings

  val soundprefix = "MMMPS"


  def playClientSound(soundname: String, volume: Float) {
    try {
      if ((FMLCommonHandler.instance.getEffectiveSide eq Side.CLIENT) && (Config.useSounds())) {
        val pitch: Float = 1.0f
      mcsound.playSoundFX(soundname, volume, pitch)
    }
    } catch {
      case e: NullPointerException => MuseLogger.logDebug("No Soundsystem")
    }
  }

  def makeSoundString(player: EntityPlayer, soundname: String): String = soundprefix + player.username + soundname

  def playerSound(player: EntityPlayer, soundname: String, volume: Float, pitch: Float = 1.0f, continuous: Boolean = true) {
    try {
      if ((FMLCommonHandler.instance.getEffectiveSide eq Side.CLIENT) && (Config.useSounds())) {
        val pitch: Float = 1.0f
      val unknownflag = true
      val soundid = makeSoundString(player, soundname)
      if (!soundsystem.playing(soundid)) {
        val soundfile = getSoundPoolEntry(soundname)
        val amp: Float = 16.0F * Math.max(1, volume)
        soundsystem.newSource(unknownflag, soundid, soundfile.soundUrl, soundfile.soundName, false, player.posX.toFloat, player.posY.toFloat, player.posZ.toFloat, 2, amp)
        soundsystem.setLooping(soundid, continuous)
        soundsystem.play(soundid)
      }
      soundsystem.setPitch(soundid, pitch)
      soundsystem.setPosition(soundid, player.posX.toFloat, player.posY.toFloat, player.posZ.toFloat)
      soundsystem.setVolume(soundid, Math.min(volume, 1) * this.options.soundVolume)
      soundsystem.setVelocity(soundid, player.motionX.toFloat, player.motionY.toFloat, player.motionZ.toFloat)
    }
    } catch {
      case e: NullPointerException => MuseLogger.logDebug("No Soundsystem")
    }
  }

  def stopPlayerSound(player: EntityPlayer, soundname: String) {
    try {
      if ((FMLCommonHandler.instance.getEffectiveSide eq Side.CLIENT) && (Config.useSounds())) {
        val soundid = makeSoundString(player, soundname)
      val vol = soundsystem.getVolume(soundid) - 0.1f
      if (vol > 0) {
        soundsystem.setVolume(soundid, vol)
      } else {
        soundsystem.stop(makeSoundString(player, soundname))
      }
    }} catch {
      case e: NullPointerException => MuseLogger.logDebug("No Soundsystem")
    }
  }

  def getSoundPoolEntry(s: String) = mcsound.soundPoolSounds.getRandomSoundFromSoundPool(s)
}

File: src/main/java/net/machinemuse/numina/client/sound/Musique.java
package net.machinemuse.numina.client.sound;

import net.machinemuse.numina.config.NuminaConfig;
import net.machinemuse.numina.basemod.MuseLogger;
import net.minecraft.client.Minecraft;
import net.minecraft.client.audio.PositionedSoundRecord;
import net.minecraft.client.audio.SoundHandler;
import net.minecraft.client.settings.GameSettings;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.SoundCategory;
import net.minecraft.util.SoundEvent;
import net.minecraft.util.math.BlockPos;
import net.minecraftforge.fml.common.FMLCommonHandler;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

import java.util.HashMap;

/**
 * Created by Claire on 8/27/2015.
 * <p>
 * Ported to Java by lehjr on 10/22/16.
 */
@SideOnly(Side.CLIENT)
public class Musique {
    private static HashMap<String, MovingSoundPlayer> soundMap = new HashMap<>();

    public static SoundHandler mcsound() {
        return Minecraft.getMinecraft().getSoundHandler();
    }

    public static void playClientSound(SoundEvent soundEvt, SoundCategory categoryIn, float volumeIn, BlockPos posIn) {
        if ((FMLCommonHandler.instance().getEffectiveSide() == Side.CLIENT) && NuminaConfig.useSounds()) {
            BlockPos pos = (posIn != null) ? posIn : Minecraft.getMinecraft().player.getPosition();

            // creates a sound
            PositionedSoundRecord sound = new PositionedSoundRecord(soundEvt, categoryIn, volumeIn, 1.0F, pos);
            mcsound().playSound(sound);
        }
    }

    public static String makeSoundString(EntityPlayer player, SoundEvent soundEvt) {
                return makeSoundString(player, soundEvt.getSoundName());
    }

    public static String makeSoundString(EntityPlayer player, ResourceLocation soundname) {
        return player.getUniqueID().toString() + soundname;
    }

    public static void playerSound(EntityPlayer player, ResourceLocation location, SoundCategory categoryIn, float volume, Float pitch, Boolean continuous) {
        SoundEvent soundEvent = SoundEvent.REGISTRY.getObject(location);
        if (soundEvent != null) {
//            System.out.println(soundEvent.getSoundName().toString());
            playerSound(player, soundEvent, categoryIn, volume, pitch, continuous);
        } else {
            soundEvent = new SoundEvent(location);
            if (soundEvent != null)
                playerSound(player, soundEvent, categoryIn, volume, pitch, continuous);
            else
                MuseLogger.logError("Sound event not found for " + location.toString());
        }
    }

    public static void playerSound(EntityPlayer player, SoundEvent soundEvt, SoundCategory categoryIn, float volume, Float pitch, Boolean continuous) {
        pitch = (pitch != null) ? pitch : 1.0F;
        continuous = (continuous != null) ? continuous : true;
        if ((FMLCommonHandler.instance().getEffectiveSide() == Side.CLIENT) && NuminaConfig.useSounds() && soundEvt != null) {
            String soundID = makeSoundString(player, soundEvt);
            MovingSoundPlayer sound = soundMap.get(soundID);

            if (sound != null && (sound.isDonePlaying() || !sound.canRepeat())) {
                stopPlayerSound(player, soundEvt);
                sound = null;
            }
            if (sound != null) {
                sound.updateVolume(volume).updatePitch(pitch).updateRepeat(continuous);
            } else {
//                MuseLogger.logDebug("New sound: " + soundEvt.getSoundName());
                MovingSoundPlayer newsound = new MovingSoundPlayer(soundEvt, categoryIn, player, volume * 2.0f, pitch, continuous);
                mcsound().playSound(newsound);
                soundMap.put(soundID, newsound);
            }
        }
    }

    public static void stopPlayerSound(EntityPlayer player, ResourceLocation location) {
        SoundEvent soundEvent = SoundEvent.REGISTRY.getObject(location);
        if (soundEvent != null) {
//            System.out.println(soundEvent.getSoundName().toString());
            stopPlayerSound(player, soundEvent);
        } else {
            soundEvent = new SoundEvent(location);
            if (soundEvent != null)
                stopPlayerSound(player, soundEvent);
            else
                MuseLogger.logError("Sound event not found for " + location.toString());
        }
    }

    public static void stopPlayerSound(EntityPlayer player, SoundEvent soundEvt) {
        if ((FMLCommonHandler.instance().getEffectiveSide() == Side.CLIENT) && NuminaConfig.useSounds()) {
            String soundID = makeSoundString(player, soundEvt);
            MovingSoundPlayer sound = soundMap.get(soundID);
            if (sound != null) {
                sound.stopPlaying();
                mcsound().stopSound(sound);
            }
            soundMap.remove(soundID);
//             MuseLogger.logDebug("Sound stopped: " + soundEvt.getSoundName());
        }
    }

    public GameSettings options() {
        return Minecraft.getMinecraft().gameSettings;
    }
}